include/picongpu/particles/collision/binary/RelativisticBinaryCollision.hpp

```diff
@@ -0,0 +1,513 @@
+
+                            //                            PMACC_DEVICE_ASSERT_MSG(
+                            //                                math::sqrt(pmacc::math::abs2(labMomentum0)) / gamma0 /
+                            //                                mass0 < c, "labVelocity0: %e",
+                            //                                sqrt(pmacc::math::abs2(labMomentum0)) / gamma0 / mass0);
+                            //                            PMACC_DEVICE_ASSERT_MSG(
+                            //                                math::sqrt(pmacc::math::abs2(labMomentum1)) / gamma1 /
+                            //                                mass1 < c, "labVelocity1: %e",
+                            //                                sqrt(pmacc::math::abs2(labMomentum1)) / gamma1 / mass1);
+                            // [Perez 2012] (1)
+                            // todo:: maybe this also with doubles?
+                            float3_COLL const comsVelocity
+                                = (labMomentum0 + labMomentum1) / (mass0 * gamma0 + mass1 * gamma1);
+                            float_COLL const comsVelocityAbs2 = pmacc::math::abs2(comsVelocity);
+                            float3_COLL comsMomentum0;
+                            float_COLL gammaComs, factorA, coeff0, coeff1;
+
+                            if(comsVelocityAbs2 != 0.0_COLL)
+                            {
+                                // written as (1-v)(1+v) rather than (1-v^2) for better performance when v close to c
+                                float_COLL const comsVelocityAbs = math::sqrt(comsVelocityAbs2);
+                                gammaComs = 1.0_COLL
+                                    / math::sqrt((1.0_COLL - comsVelocityAbs / c) * (1.0_COLL + comsVelocityAbs / c));
+                                // used later for comsToLab:
+                                factorA = (gammaComs - 1.0_COLL) / comsVelocityAbs2;
+
+                                // Stared gamma times mass, from [Perez 2012].
+                                coeff0 = coeff(labMomentum0, mass0, gamma0, gammaComs, comsVelocity);
+                                // PMACC_DEVICE_ASSERT(coeff0 > 0.0_COLL);
+                                // gamma^* . mass
+                                // if(coeff0 == 0.0_COLL) return;
+                                coeff1 = coeff(labMomentum1, mass1, gamma1, gammaComs, comsVelocity);
+                                // if(coeff1 == 0.0_COLL) return;
+                                // PMACC_DEVICE_ASSERT(coeff1 > 0.0_COLL);
+                                // (2) in [Perez 2012]
+                                comsMomentum0
+                                    = labToComs(labMomentum0, mass0, gamma0, gammaComs, factorA, comsVelocity);
+                            }
+                            else
+                            {
+                                // From smilei implementation, for v_coms^2 = 0
```

Comment:
Do we keep this reference?

Reply:
We could. I actually never verified it, just copied it from `smilei`. It's not in the referenced paper so that way we know where it comes from.

Reply:
Okay - let's see what the others think (in above question regarding smilei).

Reply:
You could still ask them quickly in their user chat if you want to know where it comes from.
If you know it, it can be explained with a physics reason (preferred).
If not, you could reference it more like: ``// According to Smilei version <nr.>, <file>, <line>``

Reply:
@PrometheusPi @n01r 
This is just the case when COMS frame is the same as the lab frame. Normally `factorA = (gammaComs - 1.0_COLL) / comsVelocityAbs2`. The `comsVelocity --> 0` limit for `factorA` is 1/(2c^2) which is 0.5 for c=1. 
I changed line 348 to `factorA = 1.0_COLL/ (2.0_COLL * c * c);` and removed `smilei` from the comment.